Add Part 2 Chapter 3 (#33)
ISSOtm/gb-asm-tutorial: src/part2/input.md

# Input

We have the building blocks of a game here, but we're still lacking player input.
A game that plays itself isn't very much fun, so let's fix that.

> Paste this code below your `Main` loop. This is another function, and like `Memcpy` can be executed from different places using the `call` opcode.

Reviewer: ```suggestion
Paste this code below your `Main` loop.
Like `Memcpy`, this is a function that can be reused from different places, using the `call` instruction.
```

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/input/main.asm:input-routine}}
{{#include ../../unbricked/input/main.asm:input-routine}}
```

> 

Reviewer: ```suggestion

Unfortunately, reading input on the Game Boy is fairly involved (as you can see!), and it would be quite difficult to explain what this function does right now.
So, I ask that you make an exception, and trust me that this function *does* read input.
Alright? Good!

```

The "Input Variables" section is missing, though.

> Now that we know how to use functions, lets call the `UpdateKeys` function in our main loop to read user input.

Reviewer: ```suggestion
Now that we know how to use functions, let's call the `UpdateKeys` function in our main loop to read user input.
```

> `UpdateKeys` sets `wCurKeys`, which we'll read from after calling it.

Reviewer: ```suggestion
`UpdateKeys` writes the held buttons to a location in memory that we called `wCurKeys`, which we can read from after the function returns.
```

> Because of this we only need to call `UpdateKeys` once per frame; calling it more than once is slow and causes input to be inconsistent.

Reviewer: ```suggestion
Because of this, we only need to call `UpdateKeys` once per frame.

This is good, because not only is it faster to just read the inputs last read, it also means that we will always act on the same inputs, even if the player presses or releases a button mid-frame.
```

> We're going to use the `bit` opcode, which sets the zero flag (`z`) to the value of the bit.

Reviewer: I'd rather we don't introduce a new instruction right there, if possible.

We can use this along with the `PADB` constants in hardware.inc to read a particular key.

> ```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/input/main.asm:copy_paddle}}
> {{#include ../../unbricked/input/main.asm:copy_paddle}}

Reviewer: ```suggestion
```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/input/main.asm:main}}
{{#include ../../unbricked/input/main.asm:main}}
```

```

> Now if you compile your project, you should be able to move the paddle left and right using the d-pad.

Reviewer: ```suggestion
Now, if you compile the project, you should be able to move the paddle left and right using the d-pad!!
Hooray, we have the beginnings of a game!
```

# Input

We have the building blocks of a game here, but we're still lacking player input.
A game that plays itself isn't very much fun, so let's fix that.

> Reading from the Game Boy's buttons is surprisingly *hard*, and it's outside the scope of this tutorial.

Reviewer 2: I don't think at this point it's even clear/obvious the difference from "reading the buttons" and "having an input routine". Besides, it's just adds useless information/perceived burden for no reason. We should remove this kind of warnings

Reviewer 3: agreed. although it's a small thing, it might cause some confusion amongst newcomers.

Reviewer: I think Evie is (rightfully) trying to explain why we're going to cargo-cult a function without explaining *what* it does, which contrasts quite heavily with the rest of the tutorial.

Maybe we should word it more like the above?

Reviewer 2: We're writing a tutorial proposing best practises. It's normal and completely legit that a small bias/preference on how to do things is pushed to the readers. As long as before the end we provide the information to contextualize and inform the audience, it should block us from suggesting approaches. Here, it's confusing, and it's not really clear what is the "hard thing" we're avoiding and what we are really suggesting.
I propose to remove this part and expand a bit what the input routine is accomplishing and how.
If it's really that important to you, we can add a small warning/info box at the end of the article (?) mentioning that we didn't really got access to "raw reads" of the buttons but [...]

On the top of this, the fact that then we're calling it "keys" but we're not really reading from buttons is confusing again.

Reviewer: Ok, yes, fair points.

# Input

We have the building blocks of a game here, but we're still lacking player input.
A game that plays itself isn't very much fun, so let's fix that.

Paste this code below your `Main` loop.
Like `Memcpy`, this is a function that can be reused from different places, using the `call` instruction.

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/input/main.asm:input-routine}}
{{#include ../../unbricked/input/main.asm:input-routine}}
```

Unfortunately, reading input on the Game Boy is fairly involved (as you can see!), and it would be quite difficult to explain what this function does right now.
So, I ask that you make an exception, and trust me that this function *does* read input.
Alright? Good!

Now that we know how to use functions, let's call the `UpdateKeys` function in our main loop to read user input.
`UpdateKeys` writes the held buttons to a location in memory that we called `wCurKeys`, which we can read from after the function returns.
Because of this, we only need to call `UpdateKeys` once per frame.

> This is good, because not only is it faster to just read the inputs last read, it also means that we will always act on the same inputs, even if the player presses or releases a button mid-frame.

Reviewer: I feel like the "read the inputs just read" formulation may be confusing, but I don't see a better alternative?
I considered "re-read the inputs seen at the beginning of the frame", maybe?

Reviewer: Based on external feedback, the current text should be good enough, so let's merge it.

# Input

> We have the building blocks of a game here, but we're still lacking user input.

Reviewer 2: (both ways are okay for me)

> We have the building blocks of a game here, but we're still lacking user input.

Author: Maybe someone's reading this to make PongOS for the Game Boy

A game without input isn't very much fun, so let's fix that.

Reading from the Game Boy's buttons is surprisingly *hard*, and it's outside the scope of this tutorial.
We're going to provide an input routine for this project since all we really want is to focus on input.

Paste this code below your `Main` loop. This is a *function*, that can be executed from different places using the `call` opcode.

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/input/main.asm:input-routine}}
{{#include ../../unbricked/input/main.asm:input-routine}}
```

> Before we use this function, lets practice writing our own.

Reviewer: I think asking the reader to "rewind" is going to make it harder for them.

I originally wanted to do Part II without introducing the stack, but evidently this is a bad idea; then, I think we should add an interstitial lesson (before this one) where we talk about it.
The lesson would be primarily about `call` and `ret`.

Reviewer 3: I think that would be helpful.
